src/reconcile/rec_visibility.c

```diff
@@ -603,6 +669,13 @@ __wt_rec_upd_select(WT_SESSION_IMPL *session, WT_RECONCILE *r, WT_INSERT *ins, W
         upd_saved = upd_select->upd_saved = true;
     }
 
+    /*
+     * Fixup any out of order timestamps, assert that checkpoint isn't running if we're in eviction.
+     */
+    if (__timestamp_out_of_order_fix(session, select_tw) && F_ISSET(r, WT_REC_EVICT) &&
+      F_ISSET(r, WT_REC_CHECKPOINT_RUNNING))
+        WT_ERR_PANIC(session, WT_ERROR, "Attempted to fix out of order timestamps illegally.");
```

Comment:
I can see that there is a test failure with this error being reported. Is there anything stopping a checkpoint getting started after earlier checks that try to avoid fixing up out of order timestamps?

Reply:
i.e: does the check here need to check something different to "is checkpoint running", more like "was checkpoint running when we made a decision that would be incorrect"

Reply:
> I can see that there is a test failure with this error being reported. Is there anything stopping a checkpoint getting started after earlier checks that try to avoid fixing up out of order timestamps?

That wouldn't be the case here as `WT_REC_CHECKPOINT_RUNNING` is set during `__rec_init` and not un-set at any stage. The actual issue here is that I wrongly only call the validation function if we insert updates into the history store based off whether the `upd_select->upd` is non NULL and not a tombstone. I should pass it regardless of whether its a tombstone or not. However the behaviour for when we select a globally visible tombstone is a bit different and won't trigger history store insertions, as such any subsequent checks on the update chain seem irrelevant. Suggesting there's a potential perf concern there. I don't want to add any additional checks for perceived performance issues yet until this change is correct.

Reply:
I fixed the potential perf issue too in my latest commit.

Reply:
@quchenhao I believe you added the `WT_REC_CHECKPOINT_RUNNING` flag, should we ever recheck the conditions of that flag during a round of eviction? What if the eviction can see more than a checkpoint, it starts, doesn't set `WT_REC_CHECKPOINT_RUNNING`, and then a checkpoint starts. It then will be allowed to evict OOO timestamps incorrectly I think?

Reply:
@luke-pearson I believe the visibility of eviction and checkpoint are fixed once they got their snapshot so I think it is not required to recheck the flag.

Reply:
The case you described is OK as at that time checkpoint hasn't got its snapshot so after it starts it will see more staff than eviction.

Reply:
We have already done the check before so I think no need to check `&& F_ISSET(r, WT_REC_EVICT) &&
      F_ISSET(r, WT_REC_CHECKPOINT_RUNNING)` again.